src/mps/dmrg.jl

```diff
@@ -278,7 +278,15 @@ function dmrg(PH, psi0::MPS, sweeps::Sweeps; kwargs...)
         end
 
         energy = vals[1]
-        phi::ITensor = vecs[1]
+        ## Right now there is a conversion problem in CUDA.jl where `UnifiedMemory` Arrays are being converted 
+        ## into `DeviceMemory`. This conversion line is here temporarily to fix that problem when it arises
+        ## Adapt is only called when using CUDA backend. CPU will work as implemented previously.
+        phi::ITensor = if NDTensors.iscu(tensor(phi))
+          itensor(adapt(typeof(data(phi)), tensor(vecs[1])))
```

Comment:
I think there is a function `datatype` we can use. It may only be defined for `Tensor`, if so we should define it for `ITensor`.

Also we should probably call `leaf_parenttype(datatype(phi))` to make sure it is adapting according to a leaf type and not a wrapper type.

Finally, we should define `adapt` for ITensor if it's not defined already, so then all of this could be:
```julia
adapt(leaf_parenttype(phi), vecs[1])
```
Another issue I could see happening here is that `vecs[1]` is complex-valued but `phi` is real-valued, say in the case where starting state is real but the Hamiltonian is complex. To handle that we could do:
```julia
adapt(set_eltype(leaf_parenttype(phi), eltype(vecs[1])), vecs[1])
```

Reply:
Note that if we define `NDTensors.parenttype(t::ITensor) = typeof(tensor(t))` I think `leaf_parenttype(t::ITensor)` should automatically work.